lib/src/models/axi4_bfm/axi4_compliance_checker.dart

```diff
@@ -0,0 +1,270 @@
+//           wIntf.awValid.previousValue!.toBool()) {
+//         final id = wIntf.awId?.previousValue?.toInt() ?? 0;
+//         final len = (wIntf.awLen?.previousValue?.toInt() ?? 0) + 1;
+//         writeReqMap[id] = [len, 0];
+//         lastWriteReqId = id;
+//       }
+
+//       // track write data flits
+//       if (wIntf.wValid.previousValue!.isValid &&
+//           wIntf.wValid.previousValue!.toBool()) {
+//         final id = lastWriteReqId;
+//         if (!writeReqMap.containsKey(id)) {
+//           logger.severe('There is no pending write request '
+//               'to associate with valid write data.');
+//         }
+
+//         writeReqMap[id]![1] = writeReqMap[id]![1] + 1;
+//         final len = writeReqMap[id]![0];
+//         final currCount = writeReqMap[id]![1];
+//         if (currCount > len) {
+//           logger.severe(
+//               'Sent more write data flits than indicated by the request '
+//               'with ID $id AWLEN. Expected $len but sent $currCount');
+//         } else if (currCount == len && !wIntf.wLast.previousValue!.toBool()) {
+//           logger.severe('Sent the final flit in the write data per the request '
+//               'with ID $id AWLEN but WLAST is not asserted.');
+//         }
+//       }
+//     });
+//   }
+// }
+
+/// A checker for some of the rules defined in the AXI4 interface specification.
+///
+/// This does not necessarily cover all rules defined in the spec.
+class Axi4ReadComplianceChecker extends Component {
+  /// AXI4 System Interface.
+  final Axi4SystemInterface sIntf;
+
+  /// AXI4 Read Interface.
+  final Axi4ReadInterface rIntf;
+
+  /// Creates a new compliance checker for AXI4.
+  Axi4ReadComplianceChecker(
+    this.sIntf,
+    this.rIntf, {
+    required Component parent,
+    String name = 'axi4ReadComplianceChecker',
+  }) : super(name, parent);
+
+  @override
+  Future<void> run(Phase phase) async {
+    unawaited(super.run(phase));
+
+    // wait for reset to complete
+    await sIntf.resetN.nextPosedge;
+
+    // checks to run
+    // READ REQUESTS
+    //   number of flits returned matches ARLEN if no error
+    //   if RLAST is present, asserted on the final flit only
+    //   if RID is present, every read response should match
+    //   a pending request ARID
+
+    final rLastPresent = rIntf.rLast != null;
+    final readReqMap = <int, List<int>>{};
+
+    sIntf.clk.posedge.listen((event) {
+      // capture read requests for counting
+      if (rIntf.arValid.previousValue!.isValid &&
+          rIntf.arValid.previousValue!.toBool()) {
+        final id = rIntf.arId?.previousValue?.toInt() ?? 0;
+        final len = (rIntf.arLen?.previousValue?.toInt() ?? 0) + 1;
+        readReqMap[id] = [len, 0];
+      }
+
+      // track read response flits
+      if (rIntf.rValid.previousValue!.isValid &&
+          rIntf.rValid.previousValue!.toBool()) {
+        final id = rIntf.rId?.previousValue?.toInt() ?? 0;
```

Comment:
`previousValue` will always be non-null after the first tick in the simulator -- since you're waiting for reset before you can get here, it must be non null (use `!`)